Refresh documentation (#1289)
bluesky/tiled: docs/source/getting-started/10-minutes-to-tiled.md

the entries it contains.

```{code-cell} ipython3
c
```

Let's look inside "examples"—another container.

```{code-cell} ipython3
c['examples']
```

When the container is large, we see the first several entries.

```{code-cell} ipython3
c['examples/xraydb']
```

````{tip}

These are equivalent, but the first two run faster.

```python
# fast
c['examples/xraydb']
c['examples', 'xraydb']

# slower because Python cannot "see ahead"
c['examples']['xraydb']
```

````

## Search on metadata

Every entry in Tiled has metadata, which we can access via the `metadata`
attribute. Tiled does not enforce any constraints on this by default: the
metadata may be any key–value pairs. In technical terms, it accepts arbitrary
JSON.

Let's take a peek at the first entry to get a sense of what we might
search for in this case.

```{code-cell} ipython3
x = c['examples/xraydb']
x.values().first().metadata
```

Below we'll focus queries that test that a given metadata key has a certain
value or a value in a given range. But Tiled supports many types of queries.

```{code-cell} ipython3
from tiled.queries import Key

x.search(Key('element.category') == 'nonmetal')
```

> 

Reviewer: ```suggestion
For numerical parameters, querying over a given range could also be useful:

```{code-cell} ipython3
x.search(
    16 <= Key('element.atomic_number') < 42
)
```
```

(I didn't actually test if this works!)

Queries can be chained to progressively narrow results.

```{code-cell} ipython3
x.search(
    Key('element.category') == 'nonmetal'
).search(
    Key('element.atomic_number') < 16
)
```

What other values does `element.category` take? We could answer that question
by downloading all the entries and tabulating them in Python, but it's faster
to ask Tiled to do it and just send the answer.

```{code-cell} ipython3
x.distinct('element.category', counts=True)['metadata']
```

We can stash the results in a variable and access them in various ways.

```{code-cell} ipython3
results = x.search(Key('element.category') == 'noble_gas')
print(f"Noble gases in this data set: {list(results)}")
```

We can efficiently access only the first result without downloading the
metadata for _all_ the results.

```{code-cell} ipython3

> first_result = results.values().first()

Reviewer: `results[0]` could be worth mentioning as a shorter alternative.

Reviewer: ah, never mind, it's below...

# Or, download as CSV file
c['examples/xraydb/C/edges'].export('my_table.csv')
```

We can open the files here or in any other program. They are now just files on our
local disk.

```{code-cell} ipython3
!cat my_table.csv
```

Let's download an image dataset as a PNG file.

```{code-cell} ipython3
c['examples/images/binary_blobs'].export('my_image.png')
```

Again, we can open the file here or in any other program.

```{code-cell} ipython3
:tags: [hide-input]

from IPython.display import Image

Image(filename='my_image.png')
```

Tiled tries to recognize the file format you want from the file extension, as
in `my_file.png` above. It can be also be specified explicitly using:

```{code-cell} ipython3
c['examples/images/binary_blobs'].export('my_image.png', format='image/png')
```

We can review the file formats.

```{code-cell} ipython3
c['examples/images/binary_blobs'].formats
```

Different data structures support different formats: arrays fit into different
formats than tables do.

```{code-cell} ipython3
c['examples/xraydb/C/edges'].formats
```

```{tip}
Tiled ships with support for a set of commonly-used formats, and server admins

> can add custom ones to meet their users' particular requirements.

Reviewer: ```suggestion
can add custom ones to meet their users' particular requirements. See <link> for more detail on how to do this.
```

Author: Link added.

```

## Locate data sources (e.g., files)

> Once you have identified data sets of interest, Tiled can point to where the

Reviewer: ```suggestion
Once you have identified data sets of interest in the Tiled catalog, it's easy to determine the physical location of the underlying data.
```

that fire when a new entry is created and when new data arrives.

```{code-cell} ipython3
# Collect references to active subscriptions. Otherwise, Python may
# garbage collect them, and they will never run.
subs = []

def on_child_created(update):
    "Called when a new entry is created in the container"
    print(f"New item named {update.key}")
    child_sub = update.child().subscribe()
    child_sub.new_data.add_callback(on_new_data)
    child_sub.start_in_thread(start=0)
    subs.append(child_sub)  # Keep a reference.

def on_new_data(update):
    "Called when new data is uploaded or registered for this entry"
    print(f"New data: {update.data()}")

sub = c.subscribe()
sub.child_created.add_callback(on_child_created)
sub.start_in_thread()
```

```{code-cell} ipython3
import numpy as np

ac = c.write_array(np.array([1, 2, 3]))
# Extend the array.
ac.patch(np.array([4, 5, 6]), offset=3, extend=True)
ac.patch(np.array([7, 8, 9]), offset=6, extend=True)

# Wait for subscriptions to process.
import time; time.sleep(1)
```

```{tip}
Under the hood, the pull-based methods (`read`, `search`, etc.) use HTTP REST
requests, while subscriptions use a WebSocket connection. This is why

> subscriptions can receive data as it arrives rather than polling repeatedly.

Reviewer: ```suggestion
subscriptions can receive data as it arrives rather than polling the server repeatedly.
```

Uploaded data is streamed via the WebSocket connection before it is even saved
to disk, which minimizes latency.
```

## Register data

> Tiled doesn't need to be involved when data is first written. You can register

Reviewer: "Tiled doesn't need to be involved when data is first written." -- I'm not sure the idea is coming across here, but I'm struggling to rephrase it myself.

Author: Clarified.

> datasets that already exist in storage — for example, files written directly by

Reviewer: ```suggestion
datasets that already exist in the storage — for example, files written directly by
```
? maybe not?
